adapters/freewheelssp/freewheelssp.go

```diff
@@ -0,0 +1,119 @@
+	"github.com/prebid/prebid-server/adapters"
+	"github.com/prebid/prebid-server/config"
+	"github.com/prebid/prebid-server/errortypes"
+	"github.com/prebid/prebid-server/openrtb_ext"
+	"golang.org/x/text/currency"
+	"net/http"
+)
+
+type RequestImpExt struct {
+	ZoneId int `json:"zoneId"`
+}
+
+type adapter struct {
+	endpoint string
+}
+
+func (a *adapter) MakeRequests(request *openrtb2.BidRequest, reqInfo *adapters.ExtraRequestInfo) ([]*adapters.RequestData, []error) {
+	if len(request.Imp) == 0 {
+		return nil, []error{&errortypes.BadInput{
+			Message: "No impression in bid request.",
+		}}
+	}
+
+	for i := 0; i < len(request.Imp); i++ {
+		imp := &request.Imp[i]
+		var bidderExt adapters.ExtImpBidder
+		if err := json.Unmarshal(imp.Ext, &bidderExt); err != nil {
+			return nil, []error{&errortypes.BadInput{
+				Message: fmt.Sprintf("Invalid imp.ext, %s", err.Error()),
+			}}
+		}
+
+		var impExt openrtb_ext.ImpExtFreewheelSSP
+		if err := json.Unmarshal(bidderExt.Bidder, &impExt); err != nil {
+			return nil, []error{&errortypes.BadInput{
+				Message: fmt.Sprintf("Invalid imp.ext, %s", err.Error()),
+			}}
+		}
+		requestImpExt := RequestImpExt{
+			ZoneId: impExt.ZoneId,
+		}
+
+		var err error
+		if imp.Ext, err = json.Marshal(requestImpExt); err != nil {
+			return nil, []error{&errortypes.BadInput{
+				Message: fmt.Sprintf("Unable to transfer requestImpExt to Json fomat, %s", err.Error()),
+			}}
+		}
+	}
+
+	requestJSON, err := json.Marshal(request)
+	headers := http.Header{}
+	headers["componentId"] = []string{"prebid-go"}
+
+	if err != nil {
+		return nil, []error{&errortypes.BadInput{
+			Message: fmt.Sprintf("Unable to transfer request to Json fomat, %s", err.Error()),
+		}}
+	}
+
+	requestData := &adapters.RequestData{
+		Method:  "POST",
+		Uri:     a.endpoint,
+		Body:    requestJSON,
+		Headers: headers,
+	}
+	return []*adapters.RequestData{requestData}, nil
+}
+
+func (a *adapter) MakeBids(internalRequest *openrtb2.BidRequest, externalRequest *adapters.RequestData, response *adapters.ResponseData) (*adapters.BidderResponse, []error) {
+	if response.StatusCode == http.StatusNoContent {
+		return nil, nil
+	}
+
+	if response.StatusCode != http.StatusOK {
+		return nil, []error{&errortypes.BadServerResponse{
+			Message: fmt.Sprintf("Unexpected status code: %d. Run with request.debug = 1 for more info", response.StatusCode),
+		}}
+	}
+
+	var bidResp openrtb2.BidResponse
+	if err := json.Unmarshal(response.Body, &bidResp); err != nil {
+		return nil, []error{err}
+	}
+	cur := currency.EUR.String()
+	if len(bidResp.Cur) != 0 {
+		cur = bidResp.Cur
+	}
+	bidResponse := &adapters.BidderResponse{
+		Currency: cur,
+		Bids:     []*adapters.TypedBid{},
+	}
+
+	bidType := openrtb_ext.BidTypeVideo
+
+	for _, seatBid := range bidResp.SeatBid {
+		for _, bid := range seatBid.Bid {
+			bidTmp := bid
+			bidResponse.Bids = append(bidResponse.Bids, &adapters.TypedBid{
+				Bid:     &bidTmp,
+				BidType: bidType,
+			})
+		}
+	}
```

Comment:
Please update the seat bid loop to match our latest recommended approach per https://docs.prebid.org/prebid-server/developers/add-new-bidder-go.html#adapter-code

The use of an index rather than the bid iteration variable is an important distinction.

```
for _, seatBid := range response.SeatBid {
    for i := range seatBid.Bid {
      b := &adapters.TypedBid{
        Bid:     &seatBid.Bid[i],
        BidType: bidType,
      }
      bidResponse.Bids = append(bidResponse.Bids, b)
  }
```

Reply:
got it, thanks.